### ISSUE TITLE
Cannot read property 'StyledIconBase' of undefined

### ISSUE THREAD

**Issue by Kodemon** [jacobwgillespie/styled-icons#659]
Attempting to import an icon leads to the error reported in the title of this topic:

```ts
import { PlusCircle } from "styled-icons/boxicons-solid/PlusCircle";
```

Resulting error output:

```sh
Uncaught TypeError: Cannot read property 'StyledIconBase' of undefined
URL: http://localhost:3000/app.js
Line: 75211, Column: 50
Stack: TypeError: Cannot read property 'StyledIconBase' of undefined
```

We are using the fuse-box bundler for typescript. Our tsconfig.json is as follows

```json
{
  "compilerOptions": {
    "strict": true,
    "strictPropertyInitialization": false,
    "module": "CommonJS",
    "target": "ES5",
    "sourceMap": true,
    "inlineSources": true,
    "jsx": "react",
    "baseUrl": ".",
    "importHelpers": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "paths": {
      ...
    },
    "plugins": [
      {
        "name": "typescript-styled-plugin"
      }
    ]
  }
}
```

Our fuse-box configuration:

```js
{
        homeDir: "..",
        output: "public/$name.js",
        target: "browser@es5"
}
```

We have tried various ways of making styled-icons work with fuse-box, and at this point we don't know if  its because of fuse-box itself or a miss-configuration somewhere. If anyone could shed some light on the error and a possibly remedy that would be great.

**Comment by mattgi** [comment]
Switching to use aliasing to point to the esm entry points results in a rabbit hole as some of the `../../` imports in the icons themselves cannot find their dependencies (`import { StyledIconBase } from '../../StyledIconBase';`).

We also run in to issues with icon library dependencies like `ThreeDRotation` being undefined.

**Comment by Kodemon** [comment]
Looking at the https://github.com/jacobwgillespie/styled-icons/issues/614 ticket I attempted to revert from 7.2.0 => 7.0.0 and it is not importing without errors.

**Comment by jacobwgillespie** [comment]
@Kodemon I did some digging into what's going on, and it looks like there's two different issues with FuseBox that are affecting this:

**First,** FuseBox is not resolving the `main` or `module` fields in `package.json` with extensions - for instance, the `package.json` for `PlusCircle` currently looks like this:

```
{
  "private": true,
  "main": "./PlusCircle",
  "module": "./PlusCircle.esm"
}
```

In every other bundler I've tested, it has added the `.js` extension when attempting to resolve those files, correctly finding `PlusCircle.js` and `PlusCircle.esm.js`.  FuseBox doesn't appear to do this and instead looks for a file without the extension, which doesn't exist.  I looked for an equivalent setting to Webpack's `module.extensions` option, but it doesn't look like that exists for FuseBox.

I believe I am able to explicitly add the `.js` extension to those fields and it should work with all other bundlers, so I've added that extension and released as `v4.7.2`.

**Second,** it looks like FuseBox requires specifying either `useJsNext: true` or `useJsNext: ['styled-icons']` in your configuration for it to pick up on the `modules` field in a `package.json` file, and unfortunately it looks like that setting only applies to the first `package.json` in a library and not the deeply-nested ones.

So FuseBox would resolve ES modules in `styled-icons/package.json`, but not in `styled-icons/boxicons-solid/PlusCircle/package.json`.

That breaks the ability to tree-shake Styled Icons with FuseBox, since if you import an icon, it will end up importing the entire set of all files the icon file requires.

A partial workaround for this is to import icons directly, as you've done, and to never import from an icon pack import:

```typescript
// This should be relatively okay:
import { PlusCircle } from "styled-icons/boxicons-solid/PlusCircle";

// This will bundle all Boxicon solid icons
import { PlusCircle } from "styled-icons/boxicons-solid";
```

Other bundlers like Webpack, Parcel, and Rollup do not have this restriction and are able to properly tree-shake.

---

@mattgi are you still experiencing missing icons with the latest version of Styled Icons?  Or are you saying they are only missing when using FuseBox?
